lib/ecs_deploy/instance_fluctuation_manager.rb

```diff
@@ -0,0 +1,173 @@
+      if az_to_container_instances.empty?
+        @logger.info("There are no instances to terminate.")
+        return
+      end
+
+      target_container_instances = extract_target_container_instances(decrease_count, az_to_container_instances)
+
+      threads = []
+      all_running_task_arns = []
+      target_container_instances.map(&:container_instance_arn).each_slice(MAX_DESCRIBABLE_CONTAINER_COUNT) do |arns|
+        ecs_client.update_container_instances_state(
+          cluster: @cluster,
+          container_instances: arns,
+          status: "DRAINING"
+        )
+        arns.each do |arn|
+          threads << Thread.new(arn) do |a|
+            all_running_task_arns.concat(stop_tasks(a))
+          end
+        end
+      end
+
+      threads.each(&:join)
+      ecs_client.wait_until(:tasks_stopped, cluster: @cluster, tasks: all_running_task_arns) unless all_running_task_arns.empty?
+      @logger.info("All running tasks are stopped")
+
+      instance_ids = target_container_instances.map(&:ec2_instance_id)
+      terminate_instances(instance_ids)
+      @logger.info("Succeeded in decreasing instances!")
+    end
+
+    private
+
+    def aws_params
+      {
+        access_key_id: EcsDeploy.config.access_key_id,
+        secret_access_key: EcsDeploy.config.secret_access_key,
+        region: @region,
+        logger: @logger
+      }.reject do |_key, value|
+        value.nil?
+      end
+    end
+
+    def as_client
+      @as_client ||= Aws::AutoScaling::Client.new(aws_params)
+    end
+
+    def ec2_client
+      @ec2_client ||= Aws::EC2::Client.new(aws_params)
+    end
+
+    def ecs_client
+      @ecs_client ||= Aws::ECS::Client.new(aws_params)
+    end
+
+    # Extract container instances to terminate considering AZ balance
+    def extract_target_container_instances(decrease_count, az_to_container_instances)
+      target_container_instances = []
+      decrease_count.times do
+        @logger.debug do
+          "AZ balance: #{az_to_container_instances.sort_by {|az, _| az }.map {|az, instances| [az, instances.size] }.to_h}"
+        end
+        az = az_to_container_instances.max_by {|_az, instances| instances.size }.first
+        target_container_instances << az_to_container_instances[az].pop
+      end
+      @logger.info do
+        "AZ balance: #{az_to_container_instances.sort_by {|az, _| az }.map {|az, instances| [az, instances.size] }.to_h}"
+      end
+
+      target_container_instances
+    end
+
+    def stop_tasks(arn)
+      running_task_arns = ecs_client.list_tasks(cluster: @cluster, container_instance: arn).task_arns
+      unless running_task_arns.empty?
+        running_tasks = ecs_client.describe_tasks(cluster: @cluster, tasks: running_task_arns).tasks
+        running_tasks.each do |task|
+          ecs_client.stop_task(cluster: @cluster, task: task.task_arn) if task.group.start_with?("family:")
+        end
+        ecs_client.wait_until(:tasks_stopped, cluster: @cluster, tasks: running_task_arns)
```

Comment:
Is this `wait_until`unnecessary because we call it in the below line ?

https://github.com/reproio/ecs_deploy/blob/13c165b82460908ea39a5ed13e6f901e688c9e35/lib/ecs_deploy/instance_fluctuation_manager.rb#L84

Reply:
Good catch! Thanks!

Fixed in ccc7d49